scheduler/src/cook/authorization.clj

```diff
@@ -0,0 +1,180 @@
+  it requires.
+
+  The `user` is a string username, the authenticated identity of the
+  agent that initiated the attempt.
+  
+  The `verb` is a keyword describing what the user is attempting to do:
+  one of #{ :create, :read, :update, :destroy, :access }. :access
+  implies all of the above. 
+
+  This set of verbs is not fixed per se; the exact interpretation of its 
+  values is up to the auth function.
+
+  The `object` is an implementation of the `Ownable` protocol that the
+  user is attempting to manipulate, such as a
+  `cook.mesos.api/Job`. Ownable objects are by definition owned by a
+  single specific user.
+
+  The authorization function evaluates this information to determine
+  whether the user is permitted to perform that verb on the object. If
+  so, the function returns true, else false.
+
+  The special object cook.authorization/system represents, by
+  definition, access to the Cook system itself. Auth functions should
+  only allow users with superuser / admin access to manipulate this object.
+
+  For example, a call to `(is-authorized? \"foobar\" :access cook.authorization/system)` is
+  attempting to determine whether user `\"foobar\"` is a system
+  administrator with global access to everything.
+  
+  If a particular auth system has no notion of superusers or systemwide
+  admins, it should return false.
+
+
+
+  Example authorization functions
+  -------------------------------
+
+  Two example authorization functions are provided:
+
+    * **open-auth** allows any user to do anything to any object. This
+      is suitable for development and testing.
+
+    * **configfile-admins-auth** reads a set of admin usernames from
+      the `:admins` config file. Users whose usernames are in this list are
+      admins, who are allowed to do anything to any object. Non-admins are
+       only allowed to manipulate their own objects.
+
+
+  Writing new authorization functions
+  ------------------------------------
+
+  You can easily extend the authorization system by providing your own
+  auth function that conforms to the above signature and pointing the
+  `:authorization-fn` config file key to it. 
+
+  For example, you could consult an external authorization system that
+  manages user credits or billing, or you could implement a \"teams\"
+  feature, where users are allowed to modify their teammates' jobs.
+
+"
+  (:require [cook.global-state :refer [global-state]]
+            [cook.util :refer [lazy-load-var]]
+            [clojure.tools.logging :as log]
+            [plumbing.core :refer (fnk defnk)]))
+
+
+;; The defonce is a kludge to work around
+;; https://groups.google.com/forum/#!topic/clojure/SYYYwZIrFiY -
+;; reloading a protocol causes existing objects of that type to become
+;; invalid as instances of the new protocol.
+;;
+;; If you redefine this protocol at runtime, you'll have to manually
+;; remove the defonce. Other than that, this should not cause issues.
+(defonce __ownable
+  (defprotocol Ownable
```

Comment:
I think that's not a bad idea in the long run, but overkill for this version, since the spec was just to make users allowed to do anything to all of their own owned objects.

In the future, we can easily extend the system to accomodate object types and classes by extending the protocol, if that becomes a requirement.